File: ccms/src/svc/directory/cms_data_definition_entry.c
#include "svc/directory/cms_data_definition_entry.h"
#include "data/common/cms_sub_reference.h"
#include "data/choice/cms_data_definition.h"
#include "data/scalar/cms_boolean.h"
#include "data/string/cms_visible_string.h"

int cms_data_definition_entry_encode_stream(per_stream_t *s, const cms_data_definition_entry_t *v) {
    if (!v || !v->reference || !v->definition) return CMS_ERR;
    int err;
    err = cms_sub_reference_encode_stream(s, v->reference);
    if (err) return err;
    /* cdcType OPTIONAL */
    {
        int present = (v->cdc_type_present && v->cdc_type_present->value) && v->cdc_type;
        cms_boolean_t bit = { .value = present };
        err = cms_boolean_encode_stream(s, &bit);
        if (err) return err;
        if (present) {
            err = cms_visible_string_encode_stream(s, v->cdc_type, 129);
            if (err) return err;
        }
    }
    err = cms_data_definition_encode_stream(s, v->definition);
    if (err) return err;
    return CMS_OK;
}

int cms_data_definition_entry_decode_stream(per_stream_t *s, cms_data_definition_entry_t *v) {
    if (!v || !v->reference || !v->definition) return CMS_ERR;
    int err;
    err = cms_sub_reference_decode_stream(s, v->reference);
    if (err) return err;
    /* cdcType OPTIONAL */
    {
        cms_boolean_t bit = {0};
        err = cms_boolean_decode_stream(s, &bit);
        if (err) return err;
        if (v->cdc_type_present) v->cdc_type_present->value = bit.value;
        if (bit.value) {
            if (!v->cdc_type) return CMS_ERR;
            err = cms_visible_string_decode_stream(s, v->cdc_type, 129);
            if (err) return err;
        }
    }
    err = cms_data_definition_decode_stream(s, v->definition);
    if (err) return err;
    return CMS_OK;
}

File: jcms/src/main/java/com/ysh/jcms/svc/directory/CmsDataDefinitionEntry.java
package com.ysh.jcms.svc.directory;

import com.ysh.jcms.core.CmsType;
import com.ysh.jcms.data.choice.CmsDataDefinition;
import com.ysh.jcms.data.common.CmsSubReference;
import com.ysh.jcms.data.scalar.CmsBoolean;
import com.ysh.jcms.data.string.CmsUint8Array;
import java.util.Arrays;
import java.util.List;

/**
 * DataDefinitionEntry ::= SEQUENCE {
 *     reference     [0] IMPLICIT SubReference,
 *     cdcType       [1] IMPLICIT VisibleString OPTIONAL,
 *     definition    [2] IMPLICIT DataDefinition
 * }  —  8.3.5
 *
 * Used by GetAllDataDefinition response (SEQUENCE OF DataDefinitionEntry).
 */
public class CmsDataDefinitionEntry extends CmsType {

    public CmsSubReference    reference;
    public CmsBoolean         cdcTypePresent;
    public CmsUint8Array      cdcType;        /* VisibleString OPTIONAL */
    public CmsDataDefinition  definition;

    public CmsDataDefinitionEntry() {
        this.reference      = new CmsSubReference();
        this.cdcTypePresent = new CmsBoolean();
        this.cdcType        = new CmsUint8Array();
        this.definition     = new CmsDataDefinition();
    }

    @Override
    public List<? extends CmsType> children() {
        return Arrays.asList(reference, cdcTypePresent, cdcType, definition);
    }
}
